packages/template-retail-react-app/app/contexts/index.js

```diff
@@ -27,23 +29,162 @@ import PropTypes from 'prop-types'
+ * const {root, findAndModifyFirst, itemsKey, fetchCategoryNode} = useCategories()
  *
  */
 export const CategoriesContext = React.createContext()
-export const CategoriesProvider = ({categories: initialCategories = {}, children}) => {
-    const [categories, setCategories] = useState(initialCategories)
+export const CategoriesProvider = ({treeRoot = {}, children, locale}) => {
+    const itemsKey = 'categories'
+    const DEFAULT_ROOT_CATEGORY = 'root'
+    const LOCAL_STORAGE_PREFIX = `pwa-kit-cat-`
+
+    const api = useCommerceAPI()
+    const [root, setRoot] = useState({
+        // map over the server-provided cat
+        ...treeRoot[DEFAULT_ROOT_CATEGORY],
+        [itemsKey]: treeRoot[DEFAULT_ROOT_CATEGORY]?.[itemsKey]?.map((item) => ({
+            ...item,
+            loaded: item?.loaded ?? false
+        }))
+    })
+
+    // iterates through each deep nested object and if finds object that has prop and value specified in objToFindBy
+    // argument, it replaces the current object with replacementObj, stops recursive walk and returns the whole tree.
+    // If none is found, it returns false.
+    const findAndModifyFirst = (tree, childrenKey, objToFindBy, replacementObj) => {
+        const findKeys = Object.keys(objToFindBy)
+
+        const rootIsMatch = findKeys.every((key) => isEqual(tree[key], objToFindBy[key]))
+        if (rootIsMatch) {
+            for (let prop in tree) {
+                delete tree[prop]
+            }
+            Object.assign(tree, replacementObj)
+            return tree
+        }
+
+        const findInChildren = (obj) => {
+            if (!Object.prototype.hasOwnProperty.call(obj, childrenKey)) return false
+            const children = obj[childrenKey]
+
+            const someChildIsMatch = children.some((child, idx, arr) => {
+                const childIsMatch = findKeys.every((key) => {
+                    return isEqual(child[key], objToFindBy[key])
+                })
+                if (childIsMatch) {
+                    arr[idx] = replacementObj
+                }
+                return childIsMatch
+            })
+            if (someChildIsMatch) return true
+
+            // We can't use `.some` for the search because it short-circuits
+            // and we still want to check every child.
+            const childrenModified = children.map((child) => {
+                return findInChildren(child)
+            })
+            // If any child was modified, then so was this `obj`
+            return childrenModified.some(Boolean)
+        }
+
+        const treeModified = findInChildren(tree)
+        return treeModified && tree
+    }
+
+    const fetchCategoryNode = async (id, levels = 1) => {
+        const STALE_TIME = 10000 // 10 seconds
+        // return early if there's one that is less than stale time
+        // than the stale time
+        const storageItem = window?.localStorage?.getItem(LOCAL_STORAGE_PREFIX + id + locale)
```

Comment:
```suggestion
        const storageString = window?.localStorage?.getItem(LOCAL_STORAGE_PREFIX + id + locale)
        const storageItem = storageString && JSON.parse(storageString)
```
`localStorage` [always returns a string](https://developer.mozilla.org/en-US/docs/Web/API/Storage/getItem).

Do we have any utils that handle using JSON in `localStorage` for us?